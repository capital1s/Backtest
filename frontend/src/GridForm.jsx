import React, { useState, useRef } from "react";
import { API_BASE_URL } from "./apiConfig";

function GridForm({
  setTrades,
  setHeldShares,
  setPerformance,
  ticker,
  errorMessage = "",
  successMessage = "",
  setErrorMessage,
  setSuccessMessage,
}) {
  const [form, setForm] = useState({
    shares: "",
    grid_up: "",
    grid_down: "",
    grid_increment: "",
    timeframe: "1 D",
  });
  // Remove local error/success state, use parent setters
  // const [errorMessage, setErrorMessage] = useState(null);
  // const [successMessage, setSuccessMessage] = useState(null);
  // Accept parent error/success setters
  // errorMessage and successMessage are now props, not setters
  const [isLoading, setIsLoading] = useState(false);
  const sharesRef = useRef();
  const gridUpRef = useRef();
  const gridDownRef = useRef();
  const gridIncrementRef = useRef();
  const handleChange = (e) => {
    const { name, value } = e.target;
    setForm((prev) => ({ ...prev, [name]: value }));
  };
  const handleSubmit = async (e) => {
    e.preventDefault();
    setErrorMessage && setErrorMessage(null); // Clear only error before submission
    setIsLoading(true);
    console.log("[GridForm.jsx] handleSubmit called. Form:", form);
    // Do NOT clear successMessage here; it should persist until a new success
    // Check for missing required fields
    const requiredFields = [
      "shares",
      "grid_up",
      "grid_down",
      "grid_increment",
      "timeframe",
    ];
    const missing = requiredFields.some(
      (f) => !form[f] || String(form[f]).trim() === "",
    );
    if (missing || !ticker) {
      setErrorMessage &&
        setErrorMessage("Please fill in all fields and select a ticker.");
      setIsLoading(false);
      console.log(
        "[GridForm.jsx] Validation failed: missing fields or ticker",
        form,
        ticker,
      );
      return;
    }
    // Numeric validation with specific error messages
    if (isNaN(Number(form.shares)) || Number(form.shares) <= 0) {
      setErrorMessage &&
        setErrorMessage("Invalid value for shares. Must be a positive number.");
      setIsLoading(false);
      console.log("[GridForm.jsx] Validation failed: invalid shares", form);
      return;
    }
    if (
      isNaN(Number(form.grid_increment)) ||
      Number(form.grid_increment) <= 0
    ) {
      setErrorMessage &&
        setErrorMessage(
          "Invalid value for grid increment. Must be a positive number.",
        );
      setIsLoading(false);
      console.log(
        "[GridForm.jsx] Validation failed: invalid grid increment",
        form,
      );
      return;
    }
    if (isNaN(Number(form.grid_up)) || Number(form.grid_up) < 0) {
      setErrorMessage &&
        setErrorMessage("Invalid value for grid up. Must be zero or positive.");
      setIsLoading(false);
      console.log("[GridForm.jsx] Validation failed: invalid grid up", form);
      return;
    }
    if (isNaN(Number(form.grid_down)) || Number(form.grid_down) < 0) {
      setErrorMessage &&
        setErrorMessage(
          "Invalid value for grid down. Must be zero or positive.",
        );
      setIsLoading(false);
      console.log("[GridForm.jsx] Validation failed: invalid grid down", form);
      return;
    }
    try {
      const validTickers = ["AAPL", "MSFT", "GOOG"];
      if (!validTickers.includes(ticker)) {
        setErrorMessage && setErrorMessage("Ticker: Invalid ticker selected.");
        setIsLoading(false);
        console.log("[GridForm.jsx] Invalid ticker selected:", ticker);
        return;
      }
      let fetchUrl = `${API_BASE_URL}/backtest`;
      // Use absolute URL for tests so undici can parse and MSW can intercept
      if (
        typeof globalThis.process !== "undefined" &&
        (globalThis.process.env.NODE_ENV === "test" ||
          globalThis.process.env.VITEST)
      ) {
        fetchUrl = "http://localhost/backtest";
      }
      const requestBody = JSON.stringify({
        ticker: ticker,
        shares: parseInt(form.shares),
        grid_up: parseFloat(form.grid_up),
        grid_down: parseFloat(form.grid_down),
        grid_increment: parseFloat(form.grid_increment),
        timeframe: form.timeframe,
        interval: "1min", // Add required interval field
      });
      console.log("[GridForm.jsx] Fetching:", fetchUrl, "Body:", requestBody);
      const response = await fetch(fetchUrl, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: requestBody,
      });

      // Handle both real responses and mock responses
      let responseText = "";
      try {
        responseText = response.clone
          ? await response.clone().text()
          : response.body || "";
        console.log(
          "[GridForm.jsx] fetch response status:",
          response.status,
          "body:",
          responseText,
        );
      } catch (cloneErr) {
        console.log("[GridForm.jsx] Clone error:", cloneErr);
        responseText = response.body || response.text || "";
      }

      let data = {};
      try {
        if (response.json && typeof response.json === "function") {
          data = await response.json();
        } else if (typeof responseText === "string" && responseText) {
          data = JSON.parse(responseText);
        } else {
          data = response.data || {};
        }
      } catch (jsonErr) {
        console.log("[GridForm.jsx] JSON parse error:", jsonErr);
      }
      if (!response.ok) {
        setErrorMessage &&
          setErrorMessage(`Server: ${data.error || "Server error!"}`);
        setIsLoading(false);
        // Do NOT clear successMessage here; it should persist after errors
        console.log(
          "[GridForm.jsx] Server error:",
          data.error || "Server error!",
        );
        console.log(
          "[GridForm.jsx] errorMessage after error:",
          data.error || "Server error!",
        );
        return;
      }
      // Clear error message on success
      setErrorMessage(null);
      setIsLoading(false);
      setSuccessMessage && setSuccessMessage("Backtest completed successfully");
      console.log("[GridForm.jsx] Success: Backtest completed successfully");
      console.log("[GridForm.jsx] errorMessage after success:", null);
      console.log(
        "[GridForm.jsx] successMessage after success:",
        "Backtest completed successfully",
      );
      setTrades && setTrades(data.trades || []);
      setHeldShares && setHeldShares(data.heldShares || []);
      setPerformance && setPerformance(data.performance || {});
      console.log(
        "[GridForm.jsx] setPerformance called with:",
        data.performance || {},
      );
    } catch (err) {
      setErrorMessage("Network Error: No response from server.");
      setErrorMessage &&
        setErrorMessage("Network Error: No response from server.");
      setIsLoading(false);
      // Do NOT clear successMessage here; it should persist after errors
      console.log("[GridForm.jsx] Network error:", err);
      console.log(
        "[GridForm.jsx] errorMessage after network error:",
        "Network Error: No response from server.",
      );
    }
  };
  return (
    <div>
      <section aria-labelledby="grid-form-title">
        <h2 id="grid-form-title" style={{ display: "none" }}>
          Grid Trading Backtest Form
        </h2>
        <div className="status-container">
          {isLoading && (
            <div
              aria-live="polite"
              role="status"
              style={{ color: "green", marginBottom: "1rem" }}
            >
              <span style={{ display: "flex", alignItems: "center" }}>
                <span
                  className="spinner"
                  style={{
                    width: "1.5rem",
                    height: "1.5rem",
                    border: "3px solid #ccc",
                    borderTop: "3px solid #333",
                    borderRadius: "50%",
                    animation: "spin 1s linear infinite",
                  }}
                />
                <span style={{ marginLeft: "0.5rem" }}>
                  Running backtest...
                </span>
              </span>
            </div>
          )}
          {/* Error message rendering for tests and accessibility */}
          {errorMessage && (
            <div
              role="alert"
              aria-live="assertive"
              style={{ color: "red", marginBottom: "1rem" }}
            >
              {errorMessage}
            </div>
          )}
          {/* Success message rendering for tests and accessibility */}
          {successMessage && (
            <div
              role="status"
              aria-live="polite"
              style={{ color: "green", marginBottom: "1rem" }}
            >
              {successMessage}
            </div>
          )}
        </div>
        <form
          role="form"
          onSubmit={handleSubmit}
          style={{ display: "flex", gap: "1rem", alignItems: "center" }}
          aria-describedby="grid-form-desc"
        >
          <span id="grid-form-desc" style={{ display: "none" }}>
            Enter grid trading parameters and run a backtest. All fields are
            required.
          </span>
          {/* ...existing code for form fields and button... */}
          <div role="group" aria-labelledby="ticker-display">
            <label
              id="ticker-display"
              style={{
                fontWeight: "bold",
                marginBottom: "0.5rem",
                display: "block",
              }}
            >
              Selected Ticker:{" "}
              {ticker ? (
                <strong style={{ color: "green" }}>{ticker}</strong>
              ) : (
                <em style={{ color: "#999" }}>None selected</em>
              )}
            </label>
            {!ticker && (
              <p
                style={{
                  fontSize: "0.9rem",
                  color: "#e74c3c",
                  margin: "0.5rem 0",
                  padding: "0.5rem",
                  backgroundColor: "#fdf2f2",
                  border: "1px solid #fecaca",
                  borderRadius: "4px",
                }}
              >
                ⚠️ Please select a ticker using the dropdown in the Chart
                section above before running a backtest.
              </p>
            )}
            {ticker && (
              <p
                style={{
                  fontSize: "0.9rem",
                  color: "#27ae60",
                  margin: "0.5rem 0",
                  padding: "0.5rem",
                  backgroundColor: "#f0f9f4",
                  border: "1px solid #bbf7d0",
                  borderRadius: "4px",
                }}
              >
                ✅ Ready to backtest with {ticker}
              </p>
            )}
          </div>
          <div role="group" aria-labelledby="shares-label">
            <label id="shares-label" htmlFor="shares-input">
              Shares
            </label>
            <input
              ref={sharesRef}
              id="shares-input"
              name="shares"
              type="number"
              placeholder="Shares"
              value={form.shares}
              onChange={handleChange}
              required
              min={1}
              aria-required="true"
              aria-label="Number of shares"
              aria-invalid={form.shares === "" || Number(form.shares) <= 0}
            />
          </div>
          <div role="group" aria-labelledby="grid-up-label">
            <label id="grid-up-label" htmlFor="grid-up-input">
              Grid Up
            </label>
            <input
              ref={gridUpRef}
              id="grid-up-input"
              name="grid_up"
              type="number"
              step={0.001}
              placeholder="Grid Up"
              value={form.grid_up}
              onChange={handleChange}
              required
              min={0}
              aria-required="true"
              aria-label="Grid up value"
              aria-invalid={form.grid_up === "" || Number(form.grid_up) < 0}
            />
          </div>
          <div role="group" aria-labelledby="grid-down-label">
            <label id="grid-down-label" htmlFor="grid-down-input">
              Grid Down
            </label>
            <input
              ref={gridDownRef}
              id="grid-down-input"
              name="grid_down"
              type="number"
              step={0.001}
              placeholder="Grid Down"
              value={form.grid_down}
              onChange={handleChange}
              required
              min={0}
              aria-required="true"
              aria-label="Grid down value"
              aria-invalid={form.grid_down === "" || Number(form.grid_down) < 0}
            />
          </div>
          <div role="group" aria-labelledby="grid-increment-label">
            <label id="grid-increment-label" htmlFor="grid-increment-input">
              Grid Increment
            </label>
            <input
              ref={gridIncrementRef}
              id="grid-increment-input"
              name="grid_increment"
              type="number"
              step={0.001}
              placeholder="Grid Increment"
              value={form.grid_increment}
              onChange={handleChange}
              required
              min={0.001}
              aria-required="true"
              aria-label="Grid increment value"
              aria-invalid={
                form.grid_increment === "" || Number(form.grid_increment) <= 0
              }
            />
          </div>
          <div role="group" aria-labelledby="timeframe-label">
            <label id="timeframe-label" htmlFor="timeframe-select">
              Timeframe
            </label>
            <select
              id="timeframe-select"
              name="timeframe"
              value={form.timeframe}
              onChange={handleChange}
              required
              aria-required="true"
              aria-label="Backtest timeframe"
            >
              <option value="1 D">1 Day</option>
              <option value="5 D">5 Days</option>
              <option value="1 W">1 Week</option>
              <option value="1 M">1 Month</option>
            </select>
          </div>
          <button
            type="submit"
            disabled={isLoading}
            aria-busy={isLoading ? "true" : "false"}
            aria-label="Run backtest"
          >
            {isLoading ? "Running..." : "Backtest"}
          </button>
        </form>
      </section>
    </div>
  );
}

export default GridForm;
