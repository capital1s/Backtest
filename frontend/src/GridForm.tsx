import React, {
  useState,
  useRef,
  useCallback,
  ChangeEvent,
  FormEvent,
} from "react";
import * as yup from "yup";
import { API_BASE_URL } from "./apiConfig";

export interface GridFormProps {
  setTrades: (trades: any) => void;
  setHeldShares: (shares: any) => void;
  setTicker: (ticker: string) => void;
  setTickerBlur: (blur: boolean) => void;
  setPerformance: (perf: any) => void;
  ticker: string;
  errorMessage?: string;
  successMessage?: string;
  setErrorMessage?: (msg: string) => void;
  setSuccessMessage?: (msg: string) => void;
}

interface FormState {
  ticker: string;
  shares: string;
  grid_up: string;
  grid_down: string;
  grid_increment: string;
  timeframe: string;
  [key: string]: string;
}

const schema = yup.object().shape({
  ticker: yup.string().required("Ticker is required"),
  shares: yup
    .number()
    .typeError("Shares must be a number")
    .required("Shares are required")
    .min(1, "Shares must be at least 1"),
  grid_up: yup
    .number()
    .typeError("Grid Up must be a number")
    .required("Grid Up is required")
    .min(0, "Grid Up must be at least 0"),
  grid_down: yup
    .number()
    .typeError("Grid Down must be a number")
    .required("Grid Down is required")
    .min(0, "Grid Down must be at least 0"),
  grid_increment: yup
    .number()
    .typeError("Grid Increment must be a number")
    .required("Grid Increment is required")
    .min(0.001, "Grid Increment must be at least 0.001"),
  timeframe: yup.string().required("Timeframe is required"),
});

const GridForm: React.FC<GridFormProps> = ({
  setTrades,
  setHeldShares,
  setTicker,
  setTickerBlur,
  setPerformance,
  ticker,
  errorMessage = "",
  successMessage = "",
  setErrorMessage,
  setSuccessMessage,
}) => {
  const [form, setForm] = useState<FormState>({
    ticker: ticker || "",
    shares: "",
    grid_up: "",
    grid_down: "",
    grid_increment: "",
    timeframe: "1 D",
  });
  const [loading, setLoading] = useState(false);
  // Internal message state only used if props not provided
  const [message, setMessage] = useState<{
    type: "error" | "success" | "";
    text: string;
  }>({ type: "", text: "" });
  const tickerRef = useRef<HTMLInputElement>(null);
  const sharesRef = useRef<HTMLInputElement>(null);
  const gridUpRef = useRef<HTMLInputElement>(null);
  const gridDownRef = useRef<HTMLInputElement>(null);
  const gridIncrementRef = useRef<HTMLInputElement>(null);

  const handleChange = useCallback(
    (e: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
      const { name, value } = e.target;
      setForm((prev) => ({
        ...prev,
        [name]: name === "timeframe" || name === "ticker" ? value : value,
      }));
    },
    [],
  );

  const handleSubmit = useCallback(
    async (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      let validationFailed = false;
      try {
        await schema.validate(form, { abortEarly: false });
      } catch (validationError: any) {
        validationFailed = true;
        const errorText =
          validationError.inner && validationError.inner.length > 0
            ? validationError.inner.map((err: any) => err.message).join("; ")
            : validationError.message || "Please fill in all fields";
        if (setErrorMessage) setErrorMessage(errorText);
        else setMessage({ type: "error", text: errorText });
      }
      if (validationFailed) return;
      const payload = {
        ticker: form.ticker,
        shares: Number(form.shares),
        grid_up: Number(Number(form.grid_up).toFixed(3)),
        grid_down: Number(Number(form.grid_down).toFixed(3)),
        grid_increment: Number(Number(form.grid_increment).toFixed(3)),
        timeframe: form.timeframe,
        interval: "1 min",
      };
      setLoading(true);
      try {
        const res = await fetch(`${API_BASE_URL}/backtest`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(payload),
        });
        if (!res.ok) {
          const text = await res.text();
          if (setErrorMessage)
            setErrorMessage(`API Error: ${res.status} - ${text}`);
          else
            setMessage({
              type: "error",
              text: `API Error: ${res.status} - ${text}`,
            });
          setLoading(false);
          return;
        }
        const data = await res.json();
        if (setSuccessMessage)
          setSuccessMessage("Backtest completed successfully");
        else
          setMessage({
            type: "success",
            text: "Backtest completed successfully",
          });
        if (data) {
          setPerformance(data.performance ?? null);
          setTrades(data.trades ?? []);
          setHeldShares(data.heldShares ?? []);
        }
      } catch (err: any) {
        if (setErrorMessage)
          setErrorMessage(`Network Error: ${err.message || err}`);
        else
          setMessage({
            type: "error",
            text: `Network Error: ${err.message || err}`,
          });
      } finally {
        setLoading(false);
      }
    },
    [
      form,
      setErrorMessage,
      setSuccessMessage,
      setPerformance,
      setTrades,
      setHeldShares,
    ],
  );

  return (
    <section aria-labelledby="grid-form-title">
      <h2 id="grid-form-title" style={{ display: "none" }}>
        Grid Trading Backtest Form
      </h2>
      {/* Only one visible alert/status node for accessibility and test compliance */}
      <div className="status-container">
        {(errorMessage || message.type === "error") &&
        (errorMessage || message.text) ? (
          <div
            className="status-error"
            style={{ color: "red", marginBottom: "0.5rem" }}
            role="alert"
            aria-live="assertive"
            data-testid="error-node"
          >
            {errorMessage || message.text}
          </div>
        ) : (successMessage || message.type === "success") &&
          (successMessage || message.text) ? (
          <div
            className="status-success"
            style={{ color: "green", marginBottom: "0.5rem" }}
            role="status"
            aria-live="polite"
            data-testid="success-node"
          >
            {successMessage || message.text}
          </div>
        ) : null}
      </div>
      <form
        onSubmit={handleSubmit}
        style={{ display: "flex", gap: "1rem", alignItems: "center" }}
        aria-describedby="grid-form-desc"
      >
        <span id="grid-form-desc" style={{ display: "none" }}>
          Grid trading backtest form. All fields must be completed before
          submitting.
        </span>
        <div role="group" aria-labelledby="ticker-label">
          <label id="ticker-label" htmlFor="ticker-input">
            Ticker
          </label>
          <input
            ref={tickerRef}
            id="ticker-input"
            name="ticker"
            type="text"
            placeholder="Ticker"
            value={form.ticker}
            onChange={handleChange}
            required
            aria-required="true"
            aria-label="Stock ticker symbol"
            aria-invalid={!form.ticker}
          />
        </div>
        <div role="group" aria-labelledby="shares-label">
          <label id="shares-label" htmlFor="shares-input">
            Shares
          </label>
          <input
            ref={sharesRef}
            id="shares-input"
            name="shares"
            type="number"
            placeholder="Shares"
            value={form.shares}
            onChange={handleChange}
            required
            min={1}
            aria-required="true"
            aria-label="Number of shares"
            aria-invalid={!form.shares}
          />
        </div>
        <div role="group" aria-labelledby="grid-up-label">
          <label id="grid-up-label" htmlFor="grid-up-input">
            Grid Up
          </label>
          <input
            ref={gridUpRef}
            id="grid-up-input"
            name="grid_up"
            type="number"
            step={0.001}
            placeholder="Grid Up"
            value={form.grid_up}
            onChange={handleChange}
            required
            min={0}
            aria-required="true"
            aria-label="Grid up value"
            aria-invalid={!form.grid_up}
          />
        </div>
        <div role="group" aria-labelledby="grid-down-label">
          <label id="grid-down-label" htmlFor="grid-down-input">
            Grid Down
          </label>
          <input
            ref={gridDownRef}
            id="grid-down-input"
            name="grid_down"
            type="number"
            step={0.001}
            placeholder="Grid Down"
            value={form.grid_down}
            onChange={handleChange}
            required
            min={0}
            aria-required="true"
            aria-label="Grid down value"
            aria-invalid={!form.grid_down}
          />
        </div>
        <div role="group" aria-labelledby="grid-increment-label">
          <label id="grid-increment-label" htmlFor="grid-increment-input">
            Grid Increment
          </label>
          <input
            ref={gridIncrementRef}
            id="grid-increment-input"
            name="grid_increment"
            type="number"
            step={0.001}
            placeholder="Grid Increment"
            value={form.grid_increment}
            onChange={handleChange}
            required
            min={0.001}
            aria-required="true"
            aria-label="Grid increment value"
            aria-invalid={!form.grid_increment}
          />
        </div>
        <div role="group" aria-labelledby="timeframe-label">
          <label id="timeframe-label" htmlFor="timeframe-select">
            Timeframe
          </label>
          <select
            id="timeframe-select"
            name="timeframe"
            value={form.timeframe}
            onChange={handleChange}
            required
            aria-required="true"
            aria-label="Backtest timeframe"
          >
            <option value="1 D">1 Day</option>
            <option value="5 D">5 Days</option>
            <option value="1 W">1 Week</option>
            <option value="1 M">1 Month</option>
          </select>
        </div>
        <button
          type="submit"
          disabled={loading}
          aria-busy={loading ? "true" : "false"}
          aria-label="Run backtest"
        >
          {loading ? "Running..." : "Backtest"}
        </button>
      </form>
    </section>
  );
};

export default React.memo(GridForm);
