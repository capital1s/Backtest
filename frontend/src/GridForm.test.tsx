// @vitest-environment jsdom
import React from "react";
import { describe, it, expect } from "vitest";
import { beforeAll, afterAll, vi } from "vitest";
import { render } from "@testing-library/react";
// Prevent network calls during tests
beforeAll(() => {
  global.fetch = vi.fn(() => Promise.resolve({
    ok: true,
    status: 200,
    json: () => Promise.resolve({}),
    text: () => Promise.resolve("")
  } as Response));
});

afterAll(() => {
  // @ts-ignore
  global.fetch = undefined;
});
import GridForm from "./GridForm";

describe("GridForm", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(
      <GridForm
        setTrades={() => {}}
        setHeldShares={() => {}}
        setTicker={() => {}}
        setTickerBlur={() => {}}
        setPerformance={() => {}}
        ticker="AAPL"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
