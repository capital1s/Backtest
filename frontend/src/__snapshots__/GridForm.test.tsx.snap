// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GridForm > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <section
      aria-labelledby="grid-form-title"
    >
      <h2
        id="grid-form-title"
        style="display: none;"
      >
        Grid Trading Backtest Form
      </h2>
      <div
        class="status-container"
      />
      <form
        aria-describedby="grid-form-desc"
        role="form"
        style="display: flex; gap: 1rem; align-items: center;"
      >
        <span
          id="grid-form-desc"
          style="display: none;"
        >
          Enter grid trading parameters and run a backtest. All fields are required.
        </span>
        <div
          aria-labelledby="ticker-label"
          role="group"
        >
          <label
            for="ticker-select"
            id="ticker-label"
          >
            Ticker
          </label>
          <select
            aria-label="Ticker"
            aria-required="true"
            id="ticker-select"
            name="ticker"
            required=""
          >
            <option
              value=""
            >
              Select ticker
            </option>
            <option
              value="AAPL"
            >
              AAPL
            </option>
            <option
              value="MSFT"
            >
              MSFT
            </option>
            <option
              value="GOOG"
            >
              GOOG
            </option>
          </select>
        </div>
        <div
          aria-labelledby="shares-label"
          role="group"
        >
          <label
            for="shares-input"
            id="shares-label"
          >
            Shares
          </label>
          <input
            aria-invalid="true"
            aria-label="Number of shares"
            aria-required="true"
            id="shares-input"
            min="1"
            name="shares"
            placeholder="Shares"
            required=""
            type="number"
            value=""
          />
        </div>
        <div
          aria-labelledby="grid-up-label"
          role="group"
        >
          <label
            for="grid-up-input"
            id="grid-up-label"
          >
            Grid Up
          </label>
          <input
            aria-invalid="true"
            aria-label="Grid up value"
            aria-required="true"
            id="grid-up-input"
            min="0"
            name="grid_up"
            placeholder="Grid Up"
            required=""
            step="0.001"
            type="number"
            value=""
          />
        </div>
        <div
          aria-labelledby="grid-down-label"
          role="group"
        >
          <label
            for="grid-down-input"
            id="grid-down-label"
          >
            Grid Down
          </label>
          <input
            aria-invalid="true"
            aria-label="Grid down value"
            aria-required="true"
            id="grid-down-input"
            min="0"
            name="grid_down"
            placeholder="Grid Down"
            required=""
            step="0.001"
            type="number"
            value=""
          />
        </div>
        <div
          aria-labelledby="grid-increment-label"
          role="group"
        >
          <label
            for="grid-increment-input"
            id="grid-increment-label"
          >
            Grid Increment
          </label>
          <input
            aria-invalid="true"
            aria-label="Grid increment value"
            aria-required="true"
            id="grid-increment-input"
            min="0.001"
            name="grid_increment"
            placeholder="Grid Increment"
            required=""
            step="0.001"
            type="number"
            value=""
          />
        </div>
        <div
          aria-labelledby="timeframe-label"
          role="group"
        >
          <label
            for="timeframe-select"
            id="timeframe-label"
          >
            Timeframe
          </label>
          <select
            aria-label="Backtest timeframe"
            aria-required="true"
            id="timeframe-select"
            name="timeframe"
            required=""
          >
            <option
              value="1 D"
            >
              1 Day
            </option>
            <option
              value="5 D"
            >
              5 Days
            </option>
            <option
              value="1 W"
            >
              1 Week
            </option>
            <option
              value="1 M"
            >
              1 Month
            </option>
          </select>
        </div>
        <button
          aria-busy="false"
          aria-label="Run backtest"
          type="submit"
        >
          Backtest
        </button>
      </form>
    </section>
  </div>
</DocumentFragment>
`;
